src/vs/workbench/contrib/userDataSync/browser/userDataSyncAccount.ts

```diff
@@ -0,0 +1,251 @@
+
+	readonly accountProviderId: string | undefined;
+
+	private _status: AccountStatus = AccountStatus.Uninitialized;
+	get status(): AccountStatus { return this._status; }
+	private readonly _onDidChangeStatus = this._register(new Emitter<AccountStatus>());
+	readonly onDidChangeStatus = this._onDidChangeStatus.event;
+
+	private _all: IUserDataSyncAccount[] = [];
+	get all(): IUserDataSyncAccount[] { return this._all; }
+
+	get current(): IUserDataSyncAccount | undefined { return this._all.filter(({ sessionId }) => sessionId === this.currentSessionId)[0]; }
+
+	constructor(
+		@IAuthenticationService private readonly authenticationService: IAuthenticationService,
+		@IAuthenticationTokenService private readonly authenticationTokenService: IAuthenticationTokenService,
+		@IQuickInputService private readonly quickInputService: IQuickInputService,
+		@IStorageService private readonly storageService: IStorageService,
+		@IUserDataSyncEnablementService private readonly userDataSyncEnablementService: IUserDataSyncEnablementService,
+		@ITelemetryService private readonly telemetryService: ITelemetryService,
+		@ILogService private readonly logService: ILogService,
+		@IProductService productService: IProductService,
+		@IConfigurationService configurationService: IConfigurationService,
+	) {
+		super();
+		this.accountProviderId = getUserDataSyncStore(productService, configurationService)?.authenticationProviderId;
+		if (this.accountProviderId) {
+			if (authenticationService.isAuthenticationProviderRegistered(this.accountProviderId)) {
+				this.initialize();
+			} else {
+				this._register(Event.once(Event.filter(this.authenticationService.onDidRegisterAuthenticationProvider, providerId => providerId === this.accountProviderId))(() => this.initialize()));
+			}
+		}
+	}
+
+	private async initialize(): Promise<void> {
+		await this.update();
+
+		this._register(
+			Event.any(
+				Event.filter(
+					Event.any(
+						this.authenticationService.onDidRegisterAuthenticationProvider,
+						this.authenticationService.onDidUnregisterAuthenticationProvider,
+					), providerId => providerId === this.accountProviderId),
+				this.authenticationTokenService.onTokenFailed)
+				(() => this.update()));
+
+		this._register(Event.filter(this.authenticationService.onDidChangeSessions, e => e.providerId === this.accountProviderId)(({ event }) => this.onDidChangeSessions(event)));
+		this._register(this.storageService.onDidChangeStorage(e => this.onDidChangeStorage(e)));
+	}
+
+	private async update(): Promise<void> {
+
+		let status = AccountStatus.Unavailable;
+		let allAccounts: Map<string, IUserDataSyncAccount> = new Map<string, IUserDataSyncAccount>();
+
+		if (this.accountProviderId) {
+
+			let currentAccount: IUserDataSyncAccount | null = null;
+			let currentSession: AuthenticationSession | undefined = undefined;
+
+			if (this.currentSessionId) {
+				const sessions = await this.authenticationService.getSessions(this.accountProviderId) || [];
```

Comment:
I think allAccounts should still be set even if there isn't a current session. For example, if a user signs into an MSA with another extension before setting sync, we would want that account to be an available option

Reply:
Good catch. 👍